app/code/community/Twispay/Tpay/controllers/PaymentController.php

```diff
@@ -1,362 +1,280 @@
-    	$this->recursiveKeySort($data);
-    
-    
-    	// Build an encoded HTTP query string from the data
-    	$query = http_build_query($data);
-    
-    	// Encrypt the query string with SHA-512 algorithm
-    	$encoded = hash_hmac('sha512', $query, $apiKey, true);
-    
-    	$checksum = base64_encode($encoded);
-    
-    	return $checksum;
+  }
+
+
+  /**
+   * Function that processes the IPN (Instant Payment Notification) message of the server.
+   */
+  public function serverAction(){
+    Mage::Log(__FUNCTION__ . ': Process the IPN response of the Twispay server.', Zend_Log::NOTICE, $this->logFileName);
+
+    $storeId = Mage::app()->getStore()->getStoreId();
+    /* Read the configuration values. */
+    $liveMode = Mage::getStoreConfig('payment/tpay/liveMode', $storeId);
+    Mage::Log(__FUNCTION__ . ': storeId=' . $storeId . ' liveMode=' . $liveMode, Zend_Log::DEBUG, $this->logFileName);
+
+    /* Check if the plugin is set to the live mode. */
+    $apiKey = (1 == $liveMode) ? (Mage::getStoreConfig('payment/tpay/liveApiKey', $storeId)) : (Mage::getStoreConfig('payment/tpay/stagingApiKey', $storeId));
+    Mage::Log(__FUNCTION__ . ': apiKey=' . $apiKey, Zend_Log::DEBUG, $this->logFileName);
+
+    /* Check if we received a response. */
+    if( (FALSE == isset($_POST['opensslResult'])) && (FALSE == isset($_POST['result'])) ) {
+      Mage::log(__FUNCTION__ . ": NULL response recived.", Zend_Log::ERR, $this->logFileName, /*forceLog*/TRUE);
+      $this->_redirect('checkout/onepage/failure', array ('_secure' => true));
     }
-    
-
-    /**
-     * Sort the array based on the keys
-     * @param array $data
-     */
-    private function recursiveKeySort(array &$data) {
-    	ksort($data, SORT_STRING);
-    	foreach ($data as $key => $value) {
-    		if (is_array($value)) {
-    			$this->recursiveKeySort($data[$key]);
-    		}
-    	}
+
+    /* Get the server response. */
+    $response = (isset($_POST['opensslResult'])) ? ($_POST['opensslResult']) : ($_POST['result']);
+
+    /* Decrypt the response. */
+    $decrypted = Mage::helper('tpay')->twispay_tw_decrypt_message(/*tw_encryptedResponse*/$response, /*secretKey*/$apiKey);
+
+    if(FALSE == $decrypted){
+      Mage::log(__FUNCTION__ . ": Failed to decript the response.", Zend_Log::ERR, $this->logFileName, /*forceLog*/TRUE);
+      $this->_redirect('checkout/onepage/failure', array ('_secure' => true));
     }
 
-} 
+    /* Validate the decripted response. */
+    $orderValidation = Mage::helper('tpay')->twispay_tw_checkValidation($decrypted);
+
+    if(TRUE !== $orderValidation){
+      Mage::log(__FUNCTION__ . ": Failed to validate the response.", Zend_Log::ERR, $this->logFileName, /*forceLog*/TRUE);
+      $this->_redirect('checkout/onepage/failure', array ('_secure' => true));
+    }
+
+    /* Extract the order. */
+    $orderId = explode('_', $decrypted['externalOrderId'])[0];
+    $order = Mage::getModel('sales/order')->loadByIncrementId($orderId);
```

Comment:
Variable `$order` is not used, is there any reason to call `loadByIncrementId`?

Reply:
I removed the line.